src/System.cc

```diff
@@ -451,29 +466,22 @@ void System::Reset()
 
 void System::Shutdown()
 {
-    mpLocalMapper->RequestFinish();
-    mpLoopCloser->RequestFinish();
     if(mpViewer)
-    {
+    {    
         mpViewer->RequestFinish();
-        while(!mpViewer->isFinished())
-        {
-            std::this_thread::sleep_for(std::chrono::microseconds(5000));
-        }
+        mptViewer->join();
     }
 
-    // Wait until all thread have effectively stopped
-    while(!mpLocalMapper->isFinished() || !mpLoopCloser->isFinished() || mpLoopCloser->isRunningGBA())
```

Comment:
the code is cleaner with your changes, but you manage the case `mpLoopCloser->isRunningGBA() `?

Reply:
`mpLoopCloser->isRunningGBA()` is here to wait a thread create by `mpLoopCloser` (named `mpThreadGBA`), this thread execute `RunGlobalBundleAdjustment`.
The pipeline work like that: The main loop of `mpLoopCloser` call `CorrectLoop` in each iteration, then `CorrectLoop` check if `mpThreadGBA` is running  (by calling `isRunningGBA()`) , and if it's true, `mpThreadGBA` is detached in order to create another thread with the same function (`RunGlobalBundleAdjustment`).
When the main loop is finished, I've added a join on this thread to wait in `mpLoopCloser` instead of wait in the system.

Reply:
ok 👍  :spaghetti: